audition/__init__.py

```diff
@@ -1 +1,116 @@
+            distance_from_best_table=self.distance_from_best_table,
+            train_end_times=train_end_times,
+            initial_model_group_ids=model_group_ids,
+            initial_metric_filters=initial_metric_filters
+        )
+        self.model_group_performance_plotter = ModelGroupPerformancePlotter(self.distance_from_best_table)
+
+        self.selection_rule_picker = SelectionRulePicker(self.distance_from_best_table)
+        self.selection_rule_plotter = SelectionRulePlotter(self.selection_rule_picker)
+        self.selection_rule_performance_plotter = SelectionRulePerformancePlotter(self.selection_rule_picker)
+
+        self.distance_from_best_table.create_and_populate(
+            model_group_ids,
+            self.train_end_times,
+            self.metrics
+        )
+
+    @property
+    def metrics(self):
+        return [
+            {'metric': f['metric'], 'parameter': f['parameter']}
+            for f in self.metric_filters
+        ]
+
+    @property
+    def thresholded_model_group_ids(self):
+        return self.model_group_thresholder.model_group_ids
+
+    @property
+    def selection_rule_model_group_ids(self):
+        model_group_ids = set()
+        for metric_definition in self.metrics:
+            for selection_rule in self.selection_rules:
+                model_group_ids |= set([
+                    result['model_group_id']
+                    for result in self.selection_rule_picker.results_for_rule(
+                        bound_selection_rule=selection_rule,
+                        model_group_ids=self.thresholded_model_group_ids,
+                        train_end_times=self.train_end_times,
+                        regret_metric=metric_definition['metric'],
+                        regret_parameter=metric_definition['parameter'],
+                    )
+                ])
+        return list(model_group_ids)
+
+    def plot_model_groups(self):
+        self.best_distance_plotter.plot_all_best_dist(
+            self.metrics,
+            self.thresholded_model_group_ids,
+            self.train_end_times
+        )
+        self.model_group_performance_plotter.plot_all(
+            metric_filters=self.metric_filters,
+            model_group_ids=self.thresholded_model_group_ids,
+            train_end_times=self.train_end_times
+        )
+
+    def update_metric_filters(self, new_filters, plot=True):
+        self.model_group_thresholder.update_filters(new_filters)
+        if plot:
+            self.plot_model_groups()
+
+    def plot_selection_rules(self):
+        for metric_definition in self.metrics:
+            common_kwargs = dict(
+                bound_selection_rules=self.selection_rules,
+                regret_metric=metric_definition['metric'],
+                regret_parameter=metric_definition['parameter'],
+                model_group_ids=self.thresholded_model_group_ids,
+                train_end_times=self.train_end_times,
```

Comment:
You won't be able to calculate regrets for the very last time frame, so I think this probably needs to be `self.train_end_times[:-1]` (assuming, of course, that `self.train_end_times` is ordered, which we should do in the constructor.